### DISCUSSION TITLE
Rich for library authors

### DISCUSSION THREAD

**Discussion by PhilipVinc** [Textualize/rich#1924] [Q&A]
Hi, 

I develop a large library for machine-learning/quantum physics based on Jax, and I like rich and would like to integrate it in our framework.

There's 1 thing I'd like to do in particular: use colours in the `repr` of a structure to better signal our users when a diagnostic is bad or good. see the screenshot for example 

<img width="390" alt="Schermata 2022-02-03 alle 12 56 47" src="https://user-images.githubusercontent.com/2407108/152692249-10f5af35-2883-4925-b632-f465f7e460f3.png">

The structure is a custom data class defined in our codebase. I would like to have the structure be printed by rich when the REPL displays it, but I don't want to force our users to `install()` `rich.pretty`.

To make it display nicely I can:
 - `from rich.pretty import install; install()` when users import our libraries, and then define a `__rich__` method on our structure. I think I cannot do this as it is beyond of our scope to activate rich for users.
 - inherit from `JupyterMixin` and define `__rich__` but that is not a great solution because it confuses our class hierarchy. Also `JupyterMixin` is not part of the documented API. Also, this will only work in ipython.
 - define `_ipython_display_(self): rich.print(self)` and `__rich__`, but defining that is not recommended so... also not an ideal option and also would only work in ipython.
 
Are there any other options, that ideally will also work in a 'bare' python repl?

EDIT:
upon some investigating, I realised that if I define the `__rich__` method together with 
```
    def __repr__(self):
        import rich
        console = rich.get_console()
        segments = list(console.render(self, console.options))
        text = console._render_buffer(segments)
        return text
```
then the `rich` representation will be used even in python REPL! that is exactly what I need.
However is this considered API stable?

**Comment by willmcgugan** [comment]
You could define a `__repr__` on your object that would print with rich, capture the output, and return the string with ansi codes.

However, I can't recommend that. If the result is not printed in a terminal, it will look like garbage.

The most reliable way would be to use the rich display hook and define a `__rich__` method. If you don't want to ask users to do the `from rich.pretty` perhaps you can use the rich extension.

```
 %load_ext rich
```

**Reply by PhilipVinc** [reply]
doesn't `%load_ext rich` do exactly  `rich.pretty.install()`?

**Reply by PhilipVinc** [reply]
>However, I can't recommend that. If the result is not printed in a terminal, it will look like garbage.

I just tried to redirect the output from to a text file and it seems to me that the output does not look like garbage, but that ansi codes are simply stripped...

why do you think they would not be stripped?

**Comment by willmcgugan** [comment]
Rich will strip ansi codes when not writing to a terminal. You will need to
construct your Console with `force_terminal=True`.

On Sun, Feb 6, 2022 at 5:49 PM Filippo Vicentini ***@***.***>
wrote:

> However, I can't recommend that. If the result is not printed in a
> terminal, it will look like garbage.
>
> I just tried to redirect the output from to a text file and it seems to me
> that the output does not look like garbage, but that ansi codes are simply
> stripped...
>
> why do you think they would not be stripped?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Textualize/rich/discussions/1924#discussioncomment-2120803>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEHLAJ23YFSDDA7GK33Z4TUZ2YA5ANCNFSM5NVSKOZQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Will McGugan
https://www.willmcgugan.com <http://www.willmcgugan.com>

**Comment by jacanchaplais** [comment]
I am having the exact same issue. I have added `__rich__` special methods, and love rich's colours on my objects, but don't want to force my users to have everything else coloured after they've imported my package by using `rich.pretty.install()`. I can't seem to find a workaround. I would really like if I could make an instance of `Console` local to my package, and then have the `__rich__` special methods hook into that, without needing a global console for the IPython terminal.

**Reply by jacanchaplais** [reply]
My source code: https://github.com/jacanchaplais/showerpipe/blob/release/0.2/showerpipe/generator.py#L303
